File: src/lib/services/audit-service.ts
import { v4 as uuidv4 } from 'uuid'
import { getCollection, COLLECTIONS } from '@/lib/db/mongodb'
import type { AuditLog, AuditAction } from '@/lib/models/types'

export class AuditService {
  static async log(params: {
    entityType: string
    entityId?: string
    action: AuditAction
    actorUserId?: string
    actorEmail?: string
    actorRole?: string
    beforeState?: Record<string, unknown>
    afterState?: Record<string, unknown>
    metadata?: Record<string, unknown>
    ipAddress?: string
    userAgent?: string
    success?: boolean
    errorMessage?: string
  }): Promise<void> {
    try {
      const logsCollection = await getCollection<AuditLog>(COLLECTIONS.AUDIT_LOGS)
      
      const log: AuditLog = {
        id: uuidv4(),
        actor_user_id: params.actorUserId || null,
        actor_email: params.actorEmail || null,
        actor_role: params.actorRole || null,
        entity_type: params.entityType,
        entity_id: params.entityId || null,
        action: params.action,
        before_state: params.beforeState || null,
        after_state: params.afterState || null,
        metadata: params.metadata || null,
        ip_address: params.ipAddress || null,
        user_agent: params.userAgent || null,
        timestamp: new Date(),
        success: params.success !== false,
        error_message: params.errorMessage || null
      }
      
      await logsCollection.insertOne(log)
    } catch (error) {
      console.error('Failed to write audit log:', error)
    }
  }
  
  static async getLogs(
    page: number = 1,
    pageSize: number = 50,
    filters?: {
      entityType?: string
      action?: AuditAction
      actorUserId?: string
      startDate?: Date
      endDate?: Date
    }
  ): Promise<{ logs: AuditLog[]; total: number }> {
    const logsCollection = await getCollection<AuditLog>(COLLECTIONS.AUDIT_LOGS)
    
    const filter: Record<string, unknown> = {}
    
    if (filters?.entityType) {
      filter.entity_type = filters.entityType
    }
    if (filters?.action) {
      filter.action = filters.action
    }
    if (filters?.actorUserId) {
      filter.actor_user_id = filters.actorUserId
    }
    if (filters?.startDate || filters?.endDate) {
      filter.timestamp = {}
      if (filters.startDate) {
        (filter.timestamp as Record<string, Date>).$gte = filters.startDate
      }
      if (filters.endDate) {
        (filter.timestamp as Record<string, Date>).$lte = filters.endDate
      }
    }
    
    const total = await logsCollection.countDocuments(filter)
    const logs = await logsCollection
      .find(filter)
      .sort({ timestamp: -1 })
      .skip((page - 1) * pageSize)
      .limit(pageSize)
      .toArray()
    
    return { logs, total }
  }
}

File: backend/services/audit_service.py
from typing import Optional, List, Tuple, Dict, Any, Union
from datetime import datetime, timezone

from db.mongodb import get_database
from models.audit import AuditLog, AuditAction
import logging

logger = logging.getLogger(__name__)

class AuditService:
    @staticmethod
    async def log(
        entity_type: str,
        action: Union[AuditAction, str],
        actor_user_id: Optional[str] = None,
        actor_email: Optional[str] = None,
        actor_role: Optional[str] = None,
        entity_id: Optional[str] = None,
        before_state: Optional[Dict[str, Any]] = None,
        after_state: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None,
        success: bool = True,
        error_message: Optional[str] = None
    ) -> AuditLog:
        """Create an audit log entry"""
        db = get_database()
        
        # Convert string action to enum if needed
        if isinstance(action, str):
            try:
                action = AuditAction(action)
            except ValueError:
                # Use UPDATE as default for unknown actions
                action = AuditAction.UPDATE
        
        # Remove sensitive data from states
        if before_state:
            before_state = AuditService._sanitize_state(before_state)
        if after_state:
            after_state = AuditService._sanitize_state(after_state)
        
        audit_log = AuditLog(
            actor_user_id=actor_user_id,
            actor_email=actor_email,
            actor_role=actor_role,
            entity_type=entity_type,
            entity_id=entity_id,
            action=action,
            before_state=before_state,
            after_state=after_state,
            metadata=metadata,
            ip_address=ip_address,
            user_agent=user_agent,
            success=success,
            error_message=error_message
        )
        
        doc = audit_log.model_dump()
        doc['timestamp'] = doc['timestamp'].isoformat()
        doc['action'] = doc['action'].value if hasattr(doc['action'], 'value') else doc['action']
        
        await db.audit_logs.insert_one(doc)
        action_str = action.value if hasattr(action, 'value') else action
        logger.info(f"Audit: {action_str} on {entity_type}/{entity_id} by {actor_email}")
        
        return audit_log
    
    @staticmethod
    def _sanitize_state(state: Dict[str, Any]) -> Dict[str, Any]:
        """Remove sensitive fields from state"""
        sensitive_fields = ['password', 'password_hash', 'token', 'refresh_token', 'access_token']
        sanitized = {}
        for key, value in state.items():
            if key.lower() in sensitive_fields:
                sanitized[key] = '[REDACTED]'
            elif isinstance(value, dict):
                sanitized[key] = AuditService._sanitize_state(value)
            else:
                sanitized[key] = value
        return sanitized
    
    @staticmethod
    async def get_logs(
        page: int = 1,
        page_size: int = 50,
        entity_type: Optional[str] = None,
        action: Optional[AuditAction] = None,
        actor_user_id: Optional[str] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None
    ) -> Tuple[List[AuditLog], int]:
        """Get paginated audit logs"""
        db = get_database()
        
        # Build query
        query = {}
        if entity_type:
            query["entity_type"] = entity_type
        if action:
            query["action"] = action.value
        if actor_user_id:
            query["actor_user_id"] = actor_user_id
        if start_date or end_date:
            query["timestamp"] = {}
            if start_date:
                query["timestamp"]["$gte"] = start_date.isoformat()
            if end_date:
                query["timestamp"]["$lte"] = end_date.isoformat()
        
        # Get total count
        total = await db.audit_logs.count_documents(query)
        
        # Get logs with pagination
        skip = (page - 1) * page_size
        cursor = db.audit_logs.find(query, {"_id": 0}).skip(skip).limit(page_size).sort("timestamp", -1)
        logs_docs = await cursor.to_list(page_size)
        
        logs = []
        for doc in logs_docs:
            if doc.get('timestamp') and isinstance(doc['timestamp'], str):
                doc['timestamp'] = datetime.fromisoformat(doc['timestamp'])
            if doc.get('action') and isinstance(doc['action'], str):
                doc['action'] = AuditAction(doc['action'])
            logs.append(AuditLog(**doc))
        
        return logs, total
